odl/solvers/functional/example_functionals.py

```diff
@@ -0,0 +1,149 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Example functions used in optimization."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+
+from odl.solvers.functional.functional import Functional
+from odl.operator.operator import Operator
+from odl.space.base_ntuples import FnBase
+from odl.space.npy_ntuples import MatVecOperator
+
+
+__all__ = ('RosenbrockFunctional',)
+
+
+class RosenbrockFunctional(Functional):
+
+    """The famous Rosenbrock funciton on ``R^n``.
+
+    Notes
+    -----
+    The functional is defined for :math:`x` in :math:`\\mathbb{R}^n`,
+    :math:`n \\geq 2`, as
+
+    .. math::
+        \sum_{i=1}^{n - 1} c (x_{i+1} - x_i^2)^2 + (1 - x_i)^2
+
+    Where :math:`c` is a constant usually set to 100 which determines how ill-
+    posed the function should be.
+    It has a minimum at :math:`x = [1, \\dots, 1]`, independent of :math:`c`.
+    """
+
+    def __init__(self, space, scale=100.0):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : `FnBase`
+            Domain of the functional.
+        scale : positive float, optional
+            The scale ``c`` in the functional, determines how ill posed the
+            functional should be.
+
+        Examples
+        --------
+        >>> import odl
+        >>> r2 = odl.rn(2)
+        >>> functional = RosenbrockFunctional(r2)
+        >>> functional([1, 1])  # optimum is 0 at [0, 0]
+        0.0
+        >>> functional([0, 1])
+        200.0
+
+        Can also be used in higher dimensions
+
+        >>> r5 = odl.rn(5)
+        >>> functional = RosenbrockFunctional(r5)
+        >>> functional([1, 1, 1, 1, 1])
+        0.0
+        """
+        self.scale = float(scale)
+        if not isinstance(space, FnBase):
+            raise ValueError('`space` must be a `FnBase`')
+        if space.size < 2:
+            raise ValueError('`space` must be at least two dimensional')
+        super().__init__(space=space, linear=False, grad_lipschitz=np.nan)
+
+    def _call(self, x):
+        """Return ``self(x)``."""
+        result = 0
+        for i in range(0, self.domain.size - 1):
+            result += self.scale * ((x[i + 1] - x[i] ** 2) ** 2 +
+                                    (x[i] - 1) ** 2)
```

Comment:
The most common case will be R2 and such, there it only makes this harder to read and slower, I doubt it's worth it. Any performance gains would be in improving the hessian to be more diagonal.